client/asset/eth/multirpc.go

```diff
@@ -0,0 +1,806 @@
+		To:        &to,
+		ChainID:   m.chainID,
+		Nonce:     nonce,
+		Gas:       txOpts.GasLimit,
+		GasFeeCap: txOpts.GasFeeCap,
+		GasTipCap: txOpts.GasTipCap,
+		Value:     txOpts.Value,
+		Data:      data,
+	}), m.chainID)
+
+	if err != nil {
+		return nil, fmt.Errorf("signing error: %v", err)
+	}
+
+	return tx, m.sendSignedTransaction(ctx, tx)
+}
+
+func (m *multiRPCClient) signData(data []byte) (sig, pubKey []byte, err error) {
+	return signData(m.creds, data)
+}
+
+func (m *multiRPCClient) syncProgress(ctx context.Context) (prog *ethereum.SyncProgress, err error) {
+	return prog, m.withAny(func(p *provider) error {
+		s, err := p.cl().SyncProgress(ctx)
+		if err != nil {
+			return fmt.Errorf("error getting sync progress from %s: %v", p.host, err)
+		}
+		if s != nil {
+			prog = s
+			return nil
+		}
+
+		// SyncProgress will return nil both before syncing has begun and after
+		// it has finished. In order to discern when syncing has begun, check
+		// that the best header came in under MaxBlockInterval.
+		bh, err := p.cl().HeaderByNumber(ctx, nil /* latest */)
+		if err != nil {
+			return fmt.Errorf("error getting header for nil SyncProgress resolution: %v", err)
+		}
+		// Time in the header is in seconds.
+		nowInSecs := time.Now().Unix() / 1000
+		timeDiff := nowInSecs - int64(bh.Time)
+
+		if timeDiff < dexeth.MaxBlockInterval {
+			// Consider this synced.
+			prog = &ethereum.SyncProgress{
+				CurrentBlock: bh.Number.Uint64(),
+				HighestBlock: bh.Number.Uint64(),
+			}
+			return nil
+		}
+		// Not synced.
+		prog = &ethereum.SyncProgress{}
+		return nil
+	})
+}
+
+func (m *multiRPCClient) transactionConfirmations(ctx context.Context, txHash common.Hash) (confs uint32, err error) {
+	var r *types.Receipt
+	var tip *types.Header
+	var notFound bool
+	if err := m.withPreferred(func(p *provider) error {
+		r, err = p.cl().TransactionReceipt(ctx, txHash)
+		if err != nil {
+			if strings.Contains(err.Error(), "not found") {
+				notFound = true
+			}
+			return err
+		}
+		tip, err = p.bestHeader(ctx, m.log)
+		return err
+	}); err != nil {
+		if notFound {
+			return 0, asset.CoinNotFoundError
+		}
+		return 0, err
+	}
+	if r.BlockNumber != nil && tip.Number != nil {
+		bigConfs := new(big.Int).Sub(tip.Number, r.BlockNumber)
+		bigConfs.Add(bigConfs, big.NewInt(1))
+		if bigConfs.IsInt64() {
+			return uint32(bigConfs.Int64()), nil
+		}
+	}
+	return 0, nil
+}
+
+func (m *multiRPCClient) txOpts(ctx context.Context, val, maxGas uint64, maxFeeRate *big.Int) (*bind.TransactOpts, error) {
+	baseFees, gasTipCap, err := m.currentFees(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if maxFeeRate == nil {
+		maxFeeRate = new(big.Int).Mul(baseFees, big.NewInt(2))
+	}
+
+	txOpts := newTxOpts(ctx, m.creds.addr, val, maxGas, maxFeeRate, gasTipCap)
+
+	nonce, err := m.nextNonce(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting nonce: %v", err)
+	}
+	txOpts.Nonce = new(big.Int).SetUint64(nonce)
```

Comment:
Yeah, so this is good.  Caller has the nonce mutex locked through txOpts -> send